### PR TITLE
Replace EraseNonKernel

### PR DESCRIPTION
**Pull request by AlexVlx**
This handles the LLVM part of removing usage of a whitelist from HCC. To wit,
it aims to replace the EraseNonKernel pass with a new pass which tries to 
build the transitive closure of all [[hc]] functions in a module, by way of 
DFS. Everything that is not part of the closure is subsequently removed. If 
a non-inlineable function is in the closure, this is an error by virtue of 
non-existing call support. At finalization all functions in the closure are 
marked as AlwaysInline. This is done so as to make it possible to get full 
inlining solely with the AlwaysInliner, without needing the more expensive 
full Inliner. The intended sequencing for the pass is that it runs on the 
fully linked bitcode, immediately before invoking AMDGPU / LC.

### CODE REVIEW

**Comment by yxsamliu** [comment]
LGTM. Thanks.

File: lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+#include <algorithm>
+#include <unordered_set>
+
+using namespace llvm;
+using namespace std;
+
+namespace
+{
+    class SelectAcceleratorCode : public ModulePass {
+        unordered_set<Function*> f_;
```
Comment (whchung) [inline lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp]: @AlexVlx could we have a more meaningful name for this set?

File: lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares and defines a pass which selects only code which is
+// expected to be run by an accelerator i.e. referenced directly of indirectly
```
Comment (scchan) [inline lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp]: I think there's a typo here:  "directly **or** indirectly"

**Review by whchung — CHANGES_REQUESTED** [review]
@AlexVlx The pass looks fine to me. If you could replace `f_` with a more readable name that would make the logic even more comprehensible.

**Review by scchan — APPROVED** [review]
found a typo; otherwise LGTM

File: lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp (hunk 2 above)
Comment (AlexVlx) [reply lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp]: Whoops - thank you for pointing it out, fixed!

File: lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp (hunk 1 above)
Comment (AlexVlx) [reply lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp]: Done, thanks for pointing it out.

File: lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        }
+
+        static
+        bool alwaysInline_(Function& x)
+        {
+            if (!x.hasFnAttribute(Attribute::AlwaysInline)) {
+                if (x.hasFnAttribute(Attribute::NoInline)) {
+                    x.removeFnAttr(Attribute::NoInline);
+                }
+                x.addFnAttr(Attribute::AlwaysInline);
```
Comment (whchung) [inline lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp:116]: this is where you all functions visited by HC kernels/functions with `alwaysinline` attribute.

File: lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares and defines a pass which selects only code which is
+// expected to be run by an accelerator i.e. referenced directly or indirectly
+// (through a fully inlineable call-chain) by a [[hc]] function. To support
+// subsequent processing, it also marks all identified functions as AlwaysInline
+// thus making it possible to use only the AlwaysInliner without resorting to a
+// more expensive full Inliner pass.
```
Comment (whchung) [inline lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp:15]: @AlexVlx thanks for putting comments in this pass. It answers one question i had in the PR for `hcc`.

**Review by whchung — APPROVED** [review]
LGTM

**Review by scchan — APPROVED** [review]

**Pull request merged**